lib/using_restricted_array.rb

```diff
@@ -6,58 +6,155 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of the array
+# The number of iterations of the loop increases proportionally with the size of the array
+# Space complexity: O(1) or constant since we need one variable to save the value at index.
+# The additional storage needed is independent of the length of the array.
 def length(array)
-  raise NotImplementedError
+  return if array.nil?
+
+  index = 0
+  until array[index].nil?
+    index +=1
+  end
+
+  return index
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of the array
+# The number of iterations of the loop increases proportionally with the size of the array
+# Space complexity: O(1) or constant since we need one variable to save the value at index.
+# The additional storage needed is independent of the length of the array.
 def print_array(array)
-  raise NotImplementedError
+  return if array.nil?
+
+  index = 0
+  while index < length(array)
```

Comment:
Just note that `length(array)` goes through the entire array with each iteration of the while loop.  This makes the method an O(n^2) time complexity method.  You can do better if you do this:

```suggestion
  len = length(array)
  while index < len
```